### PR TITLE
[improvement] Avoid leaks  when use computeIfAbsent in ConcurrentLongHashMap and ConcurrentOpenHashMap

### PR DESCRIPTION
**Pull request by Radiancebobo**
The PR was introduced in bookkeeper's PR：https://github.com/apache/bookkeeper/pull/3706

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java [hunk 1]
```diff
@@ -361,6 +361,9 @@ V get(long key, int keyHash) {
         }
 
         V put(long key, V value, int keyHash, boolean onlyIfAbsent, LongFunction<V> valueProvider) {
+            if (value == null) {
+                requireNonNull(valueProvider.apply(key));
```
Comment (congbobo184) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java:365]: Even if modification is required, we should only run `valueProvider.apply(key)` once. reason :
1. The two calculated values may be different
2. cost performance

**Review by congbobo184 — CHANGES_REQUESTED** [review]
It changes the default behavior of ConcurrentLongHashMap.computeIfAbsent(long key, LongFunction<V> provider) . I'm not sure if it really needs to be modified. @merlimat  hi, could you please take a look?

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java (hunk 1 above)
Comment (Radiancebobo) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java:365]: how about this ?
image

Comment (Radiancebobo) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java:365]: > 

yes, we should run valueProvider.apply(key) once.  like above,

Comment (Radiancebobo) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentLongHashMap.java:365]: I'll discuss it here（https://github.com/apache/bookkeeper/pull/3706） and sync it up after the merge, thank you for your review, @congbobo184

**Pull request closed without merging**